icu4c/source/tools/genrb/filterrb.cpp

```diff
@@ -0,0 +1,161 @@
+// © 2018 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <iostream>
+#include <stack>
+
+#include "filterrb.h"
+
+
+const char* PathFilter::kEInclusionNames[] = {
+    "INCLUDE",
+    "PARTIAL",
+    "EXCLUDE"
+};
+
+
+ResKeyPath::ResKeyPath() {}
+
+ResKeyPath::ResKeyPath(const std::string& path, UErrorCode& status) {
+    if (path.length()  < 1 || path[0] != '/') {
+        std::cerr << "genrb error: path must start with /: " << path << std::endl;
+        status = U_PARSE_ERROR;
+        return;
+    }
+    size_t i;
+    size_t j = 0;
+    while (true) {
+        i = j + 1;
+        j = path.find('/', i);
+        std::string key = path.substr(i, j - i);
+        if (key.length() == 0) {
+            // trailing /
+            break;
+        }
+        push(key);
+        if (j == std::string::npos) {
+            break;
+        }
+    }
+}
+
+void ResKeyPath::push(const std::string& key) {
+    fPath.push_back(key);
+}
+
+void ResKeyPath::pop() {
+    fPath.pop_back();
+}
+
+const std::list<std::string>& ResKeyPath::pieces() const {
+    return fPath;
+}
+
+std::ostream& operator<<(std::ostream& out, const ResKeyPath& value) {
+    if (value.pieces().size() == 0) {
+        out << "/";
+    } else for (auto& key : value.pieces()) {
+        out << "/" << key;
+    }
+    return out;
+}
+
+
+void SimpleRuleBasedPathFilter::addRule(const std::string& ruleLine, UErrorCode& status) {
+    if (ruleLine.length() < 1) {
+        std::cerr << "genrb error: filter rule too short: " << ruleLine << std::endl;
+        status = U_PARSE_ERROR;
+        return;
+    }
+    bool inclusionRule = false;
+    if (ruleLine[0] == '+') {
+        inclusionRule = true;
+    } else if (ruleLine[0] != '-') {
+        std::cerr << "genrb error: lines must start with + or -: " << ruleLine << std::endl;
+        status = U_PARSE_ERROR;
+        return;
+    }
+    ResKeyPath path(ruleLine.substr(1), status);
+    addRule(path, inclusionRule, status);
+}
+
+void SimpleRuleBasedPathFilter::addRule(const ResKeyPath& path, bool inclusionRule, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    Tree* node = &fRoot;
+    for (auto& key : path.pieces()) {
+        // note: operator[] auto-constructs default values
+        node = &node->fChildren[key];
+    }
+    if (node->fIncluded != PARTIAL) {
+        std::cerr << "genrb error: duplicate paths in filter file" << std::endl;
```

Comment:
print the path

Reply:
Done, and changed from an error to an info.  I see the following use case: you have a default list of rules for the top-level locale tree, but for one particular locale, you want to actually override some of the rules with your own custom rules.